velox/exec/Spill.h

```diff
@@ -0,0 +1,350 @@
+  // be used with seekp to rewind a read.
+  Position tellp() const override {
+    return std::make_tuple(
+        nullptr, offset_ - current_->size + current_->position);
+  }
+
+  // Positions the file so that the byte at 'position' is the first to be read.
+  void seekp(Position position) override;
+
+ private:
+  std::unique_ptr<ReadFile> input_;
+  BufferPtr buffer_;
+  const uint64_t size_;
+  // Offset of first byte not in 'buffer_'
+  uint64_t offset_ = 0;
+};
+
+// A source of SpillFileRows coming either from a file or memory.
+class SpillStream {
+ public:
+  SpillStream(RowTypePtr type, memory::MemoryPool& pool)
+      : type_(type), pool_(pool) {
+    ordinal_ = ++ordinalCounter_;
+    auto width = type_->size();
+    decoded_.resize(width);
+    for (auto i = 0; i < width; ++i) {
+      decoded_[i] = std::make_unique<DecodedVector>();
+    }
+  }
+
+  virtual ~SpillStream() = default;
+
+  virtual bool atEnd() const = 0;
+
+  virtual std::string label() const {
+    return fmt::format("{}", ordinal_);
+  }
+
+  SpillFileRow next() {
+    if (index_ == numRowsInVector_) {
+      nextBatch();
+      index_ = 0;
+      numRowsInVector_ = rowVector_->size();
+      SelectivityVector allRows(numRowsInVector_);
+      auto width = rowVector_->childrenSize();
+      for (auto i = 0; i < width; ++i) {
+        decoded_[i]->decode(*rowVector_->childAt(i), allRows);
+      }
+    }
+    return {&decoded_, rowVector_.get(), index_++};
+  }
+
+  virtual uint64_t size() const {
+    return 0;
+  }
+
+ protected:
+  // Loads the next RowVector from the backing storage, e.g. spill file or
+  // RowContainer.
+  virtual void nextBatch() = 0;
+
+  const std::shared_ptr<const RowType> type_;
+  memory::MemoryPool& pool_;
+  RowVectorPtr rowVector_;
+  // The next row to process from 'rowVector_'
+  vector_size_t index_ = 0;
+  // Number of rows in 'rowVector_'
+  vector_size_t numRowsInVector_ = 0;
+  std::vector<std::unique_ptr<DecodedVector>> decoded_;
+  // Ordinal number used for making a a label for debugging.
+  int32_t ordinal_;
+  static std::atomic<int32_t> ordinalCounter_;
+};
+
+// Represents a spill file that is first in write mode and then
+// turns into a source of SpillFileRow. Owns a file system file that
+// contains the spilled data and is live for the duration of 'this'.
+class SpillFile : public SpillStream {
+ public:
+  SpillFile(RowTypePtr type, const std::string& path, memory::MemoryPool& pool)
+      : SpillStream(type, pool), path_(path) {}
+
+  ~SpillFile();
+
+  // Returns a file for writing spilled data.
+  WriteFile& output();
+
+  // Finishes writing and flushes any unwritten data.
+  void finishWrite() {
+    VELOX_CHECK(output_);
+    fileSize_ = output_->size();
+    output_ = nullptr;
+  }
+
+  // Prepares 'this' for reading. Positions the read at the first row of
+  // content.
+  void startRead();
+
+  // Returns the file size in bytes.
```

Comment:
Perhaps, clarify that this method returns different value depending on whether it is called during writing or read.